### PR TITLE
Improve foreach loop to support iterable objects

### LINKED ISSUE #19741: Implement Iterable objects
**Description:**
According to the spec an iterable object is, 
An object can make itself be iterable as a sequence of values of type T by providing a method named __iterator which returns a value that is a subtype of Iterator<T>. In this specification, we refer to this type as Iterable<T>.

A iterator is a subtype of

```ballerina
abstract object {
       public next() returns record {| T value; |}?;
    }
```

### PR DESCRIPTION
**Pull request by gimantha**
## Purpose
Fixes #19741

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 1]
```diff
@@ -476,6 +476,9 @@ error.invalid.xml.ns.interpolation=\
 error.iterable.not.supported.collection=\
   incompatible types: ''{0}'' is not an iterable collection
 
+error.iterable.not.supported.object=\
+  incompatible types: ''{0}'' is not an iterable object
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: We can improve the error message saying why it is not iterable. There are two possibilities. 
1) __iterator function not found. 
2) __iterator function signature not matched.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -1043,6 +1045,106 @@ public void setForeachTypedBindingPatternType(BLangForeach foreachNode) {
         foreachNode.nillableResultType = nextMethodReturnType;
     }
 
+    private BUnionType getVarTypeFromIterableObject(BObjectType collectionType) {
+        BObjectTypeSymbol objectTypeSymbol = (BObjectTypeSymbol) collectionType.tsymbol;
+        Optional<BAttachedFunction> attachFunc = objectTypeSymbol.attachedFuncs.stream()
+                .filter(func -> func.funcName.value.equals("__iterator")).findFirst();
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Shall we move this into a constant?

Comment (hasithaa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: & replace the stream with a loop. :)

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -2506,6 +2506,32 @@ public void visit(BLangForeach foreach) {
...
 
+    private BLangBlockStmt desugarForeachWithIteratorDef(BLangForeach foreach,
+                                                         BLangSimpleVariableDef dataVariableDefinition,
+                                                         BVarSymbol collectionSymbol,
+                                                         BInvokableSymbol iteratorInvokableSymbol,
+                                                         boolean isIteratorFuncFromLangLib) {
+        BLangBlockStmt blockNode;
+        BLangSimpleVariableDef iteratorVarDef = getIteratorVariableDefinition(foreach.pos, collectionSymbol,
+                iteratorInvokableSymbol, isIteratorFuncFromLangLib);
+        blockNode = desugarForeachToWhile(foreach, iteratorVarDef);
```
Comment (irshadnilam) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Shall we inline declaration and initialization of blockNode
`BLangBlockStmt blockNode = desugarForeachToWhile(foreach, iteratorVarDef);`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -2506,6 +2506,32 @@ public void visit(BLangForeach foreach) {
...
+        blockNode.stmts.add(0, dataVariableDefinition);
+        return blockNode;
+    }
+
+    private BInvokableSymbol getIterableObjectIteratorInvokableSymbol(BVarSymbol collectionSymbol) {
+        BObjectTypeSymbol typeSymbol = (BObjectTypeSymbol) collectionSymbol.type.tsymbol;
+        //We know for sure at this point, the object symbol should have the __iterator method
```
Comment (irshadnilam) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Space before start of the comment 
`// We know for sure at this point, the object symbol should have the __iterator method`

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachIterableObjectTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+import org.ballerinalang.test.util.BRunUtil;
+import org.ballerinalang.test.util.CompileResult;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * TestCases for foreach with Maps.
+ *
+ * @since 0.96.0
```
Comment (hasithaa) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachIterableObjectTest.java]: Wrong version.

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 1 above)
Comment (gimantha) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 2 above)
Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 3 above)
Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 4 above)
Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: fixed

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachIterableObjectTest.java (hunk 5 above)
Comment (gimantha) [reply langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachIterableObjectTest.java]: fixed

**Pull request merged**